### DISCUSSION TITLE
Jinja Templating Url Swapping Scheme Context Dependant

### DISCUSSION THREAD

**Discussion by josheppinette** [encode/starlette#1981] [Potential Issue] [answered]
I am seeing the weirdest issue, when using `url_for`  inside an anchor `href` attribute, the scheme is properly rendering as `https`, but if I template that same `url_for` just directly in the html i.e. `<p>{{ url_for(...) }}</p>`, then it renders as `http://...`. 

What could be causing this?

**Comment by mxdev88** [comment]
I'm having the same issue. This causes [Mixed content](https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content) error to be raised by the browsers and makes it impossible to use `url_for` for including scripts, css, ... 

I use it within `link` and `<script>` tags:

```html
<link rel="stylesheet" href="{{ url_for('static', path='styles.css') }}">
<script src="{{ url_for('static', path='/main.js') }}"></script
```

Same issue when used in `<form>` tag:

```
<form action="{{ url_for('name') }}" method="post">
```

could this be due to this [line](https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/starlette/routing.py#L275C32-L275C32)? http scheme is a hard-coded.

**Comment by alex-oleshkevich** [comment, marked as answer]
Most probably you have an app server in front of the starlette app which does TLS termination. 
So after the termination, app server forwards HTTP request to the app and that's why you see "http" in the generated URLs.

You need to configure your server in a way that it can set `X-Forwarded-Proto` to the original value. Then add a middleware which can replace proto with `X-Forwarded-Proto`.

Similar issue - https://github.com/encode/starlette/issues/604

**Reply by josheppinette** [reply]
The issue is that `url_for` is changing behavior based on if its used inside an anchor vs regular HTML tag.

**Reply by alex-oleshkevich** [reply]
It always uses `request.url_for` in the templates, which relies on the headers: https://github.com/encode/starlette/blob/master/starlette/templating.py#L132
